38-allowlist-management.md

```diff
@@ -0,0 +1,162 @@
+  allowlist data that will be used for the agent.
+* Implements ``/v2/allowlists/`` API for Verifier to support CRD operations. Update an
+  allowlist is not included in the initial proposal but can be implemented if necessary.
+* Enhance ``tenant`` CLI to manage allowlists, when adding agent to the Verifier, user
+  specifies an existing ``allowlist`` name instead of allowlist data to complete the
+  registration.
+
+
+### User Stories (optional)
+
+
+#### Story 1
+
+A user can reuse allowlist by:
+
+* Create an allowlist record by posting policies to the Verifier.
+* Add an agent to the Verifier by providing Agent ID, IP Address and the allowlist name
+  the agent will use.
+* Add another agent running under the same environment to the Verifier with the same allowlist.
+
+#### Story 2
+
+A user can update agent with new allowlist by:
+
+* Create an allowlist record by posting new policies to the Verifier.
+* Update an agent to the Verifier by updating the allowlist name.
+* Agent will use new allowlist to verify system integrety.
+
+### Notes/Constraints/Caveats (optional)
+
+None
+
+### Risks and Mitigations
+
+None
+
+## Design Details
+
+DB Changes:
+
+The ``allowlists`` table will be created as:
+
+```
+op.create_table('allowlists',
+                sa.Column('id', sa.Integer(), nullable=False),
+                sa.Column('name', sa.String(255), nullable=False),
+                sa.Column('tpm_policy', sa.Text(), nullable=True),
+                sa.Column('vtpm_policy', sa.Text(), nullable=True),
+                sa.Column('ima_policy', sa.Text(length=429400000), nullable=True),
+                sa.PrimaryKeyConstraint('id'),
+                sa.UniqueConstraint('name', name='uniq_allowlists0name'),
+                mysql_engine='InnoDB',
+                mysql_charset='UTF8'
+                )
+```
+
+The ``verifiermain.allowlist`` column will be renamed to ``ima_policy`` in the
+``allowlists`` which provides a better presentation.
+
+The ``verifiermain`` will have a new column as foreign key to the ``allowlists`` table:
+
+```
+with op.batch_alter_table('verifiermain') as batch_op:
+    batch_op.add_column(sa.Column('allowlist_id', sa.Integer(), nullable=True))
+    batch_op.create_foreign_key('fk_verifiermain_allowlists', 'allowlists',
+                                ['allowlist_id'], ['id'])
+    batch_op.drop_column('tpm_policy')
+    batch_op.drop_column('vtpm_policy')
+    batch_op.drop_column('allowlist')
+```
+
+``tpm_policy``, ``vtpm_policy``, and ``allowlist`` will be removed from the
+``verifiermain`` table.
+
+### Test Plan
+
+Downstream testing showed that this is working as expected, will add integration
+test cases to cover this feature.
+
+
+### Upgrade / Downgrade Strategy
+
+The database migration will be handled by migration scripts.
+
+## Drawbacks
+
+Existing database records will not be migrated, after migration, agents will need
```

Comment:
Would it be feasible to actually do a migration of the existing data? I know it's still a new project and there aren't that many production systems, but I want to start shifting our mindset so that when there are we are already treating everything as it should.

For new allowlists maybe they could be created with an autogenerated name based on the uuid of the agent? This would also allow the migration to drop the old column from that table.

Reply:
Yeah, it is feasible to autogenerate allowlist name from agent id, there is overhead on the API to support both though.
For existing data migration, I am not sure there is a good way.

Reply:
I think it would be good that current usage doesn't have to change with this feature. So if I have an automated onboarding process keylime will just do the right thing behind the scenes without me needing to specify a name for my allowlist. We'll just keep multiple copies around like we do today, just in a different DB structure.